discovery_engine_core/core/src/stack/exploration.rs

```diff
@@ -0,0 +1,531 @@
+    /// Not enough cois
+    NotEnoughCois,
+}
+
+/// Configurations for the exploration stack.
+pub(crate) struct Config {
+    /// The number of candidates.
+    number_of_candidates: usize,
+    /// The maximum number of documents to keep.
+    max_selected_docs: usize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            number_of_candidates: 40,
+            max_selected_docs: 20,
+        }
+    }
+}
+
+/// Selects documents by randomization with a threshold.
+///
+/// <https://xainag.atlassian.net/wiki/spaces/M2D/pages/2376663041/Exploration+Stack#2.-Pick-a-random-article-and-delete-articles-more-similar-to-it-than-a-threshold>
+/// outlines how the selection works.
+///
+/// # Errors
+/// Fails if both positive and negative cois are empty.
+#[allow(dead_code)]
+pub(crate) fn document_selection(
+    positive_cois: &[PositiveCoi],
+    negative_cois: &[NegativeCoi],
+    documents: Vec<Document>,
+    config: &Config,
+) -> Result<Vec<Document>, Error> {
+    document_selection_with_rng(
+        positive_cois,
+        negative_cois,
+        documents,
+        config,
+        &mut rand::thread_rng(),
+    )
+}
+
+fn document_selection_with_rng<R>(
+    positive_cois: &[PositiveCoi],
+    negative_cois: &[NegativeCoi],
+    documents: Vec<Document>,
+    config: &Config,
+    rng: &mut R,
+) -> Result<Vec<Document>, Error>
+where
+    R: Rng + ?Sized,
+{
+    if positive_cois.len() + negative_cois.len() == 0 {
+        return Err(Error::NotEnoughCois);
+    }
+
+    let document_embeddings = documents
+        .iter()
+        .map(|document| document.smbert_embedding.view())
+        .collect_vec();
+
+    let pos_cois = positive_cois.iter().map(|coi| coi.point().view());
+    let neg_cois = negative_cois.iter().map(|coi| coi.point().view());
+
+    let cois = chain!(pos_cois, neg_cois).collect_vec();
+    // find_nearest_coi_for_docs can't panic because we make sure beforehand
+    // that both positive and negative cois aren't empty
+    let nearest_coi_for_docs = find_nearest_coi_for_docs(&document_embeddings, &cois);
+    let doc_similarities = pairwise_cosine_similarity(document_embeddings.into_iter());
+
+    let selected = select_by_randomization_with_threshold(
+        &doc_similarities,
+        &nearest_coi_for_docs,
+        config.number_of_candidates,
+        config.max_selected_docs,
+        rng,
+    );
+
+    Ok(retain_documents_by_indices(selected, documents))
+}
+
+fn select_by_randomization_with_threshold<R>(
+    doc_similarities: &Array2<f32>,
+    nearest_coi_for_docs: &[f32],
+    number_of_candidates: usize,
+    max_selected_docs: usize,
+    rng: &mut R,
+) -> Vec<usize>
+where
+    R: Rng + ?Sized,
+{
+    let (threshold, mut candidates) =
+        select_initial_candidates(nearest_coi_for_docs, number_of_candidates);
+
+    let mut selected = Vec::new();
+    while !candidates.is_empty() && selected.len() < max_selected_docs {
+        let chosen_doc = *candidates
```

Comment:
:wrench:  This doesn't remove a chosen candidate from the list of candidates, so candidates might be chosen multiple times.